File: Binary-Search/69-sqrtx/sqrtx.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0:
            return 0

        low, high = 1, x

        while low <= high:
            mid = low + (high - low) // 2
            sq = mid * mid

            if sq == x:
                return mid
            elif sq > x:
                high = mid - 1
            else:
                low = mid + 1

        return high  # Return the largest integer whose square is <= x

File: Binary-Search/69-sqrtx/sqrtx.cpp
class Solution {
public:
    int mySqrt(int x) {
        if(x == 0) return 0;

        int low = 1, high = x;
        int ans = -1;
        
        while(low <= high){
            int mid = low + (high - low) / 2;
            long int sq = (long int)mid * mid;

            if(sq == x){
                ans = mid;
                break;
            }
            else if(sq > x)
                high = mid - 1;
            else{
                low = mid + 1;
                ans = mid;
            }
        }

        return ans;
    }
};